### ISSUE TITLE
How to use state parameter for the OAuth2 flow in a Bolt Java implementation

### ISSUE THREAD

**Issue by mattv** [slackapi/java-slack-sdk#601]
I am implementing a slack app using Bolt Java with Spring Boot and am trying to distribute the app but am having problems getting the oAuth2 flow working. I have followed the Spring Boot example I found [here](https://slack.dev/java-slack-sdk/guides/app-distribution) but am getting and "Invalid state parameter detected" warning and the flow produces an error . I've read [here](https://github.com/slackapi/java-slack-sdk/issues/479) that Bolt Java requires the state parameter for the OAuth flows, but can't figure out how to setup my project so that state parameters are generated and properly injected into the flow and session. Any direction would be greatly appreciated!

#### The Slack SDK version

+--- com.slack.api:bolt:1.2.1
|    +--- com.slack.api:slack-api-model:1.2.1
|    +--- com.slack.api:slack-api-client:1.2.1
|    |    +--- com.slack.api:slack-api-model:1.2.1 (*)
|    \--- com.slack.api:slack-app-backend:1.2.1
|         +--- com.slack.api:slack-api-model:1.2.1 (*)
|         +--- com.slack.api:slack-api-client:1.2.1 (*)
\--- com.slack.api:bolt-servlet:1.2.1
     +--- com.slack.api:slack-api-model:1.2.1 (*)
     +--- com.slack.api:slack-api-client:1.2.1 (*)
     +--- com.slack.api:slack-app-backend:1.2.1 (*)
     \--- com.slack.api:bolt:1.2.1 (*)
+--- com.slack.api:bolt:1.2.1 (n)
\--- com.slack.api:bolt-servlet:1.2.1 (n)
+--- com.slack.api:bolt:1.2.1
|    +--- com.slack.api:slack-api-model:1.2.1
|    +--- com.slack.api:slack-api-client:1.2.1
|    |    +--- com.slack.api:slack-api-model:1.2.1 (*)
|    \--- com.slack.api:slack-app-backend:1.2.1
|         +--- com.slack.api:slack-api-model:1.2.1 (*)
|         +--- com.slack.api:slack-api-client:1.2.1 (*)
\--- com.slack.api:bolt-servlet:1.2.1
     +--- com.slack.api:slack-api-model:1.2.1 (*)
     +--- com.slack.api:slack-api-client:1.2.1 (*)
     +--- com.slack.api:slack-app-backend:1.2.1 (*)
     \--- com.slack.api:bolt:1.2.1 (*)
+--- com.slack.api:bolt:1.2.1
|    +--- com.slack.api:slack-api-model:1.2.1
|    +--- com.slack.api:slack-api-client:1.2.1
|    |    +--- com.slack.api:slack-api-model:1.2.1 (*)
|    \--- com.slack.api:slack-app-backend:1.2.1
|         +--- com.slack.api:slack-api-model:1.2.1 (*)
|         +--- com.slack.api:slack-api-client:1.2.1 (*)
\--- com.slack.api:bolt-servlet:1.2.1
     +--- com.slack.api:slack-api-model:1.2.1 (*)
     +--- com.slack.api:slack-api-client:1.2.1 (*)
     +--- com.slack.api:slack-app-backend:1.2.1 (*)
     \--- com.slack.api:bolt:1.2.1 (*)
+--- com.slack.api:bolt:1.2.1
|    +--- com.slack.api:slack-api-model:1.2.1
|    +--- com.slack.api:slack-api-client:1.2.1
|    |    +--- com.slack.api:slack-api-model:1.2.1 (*)
|    \--- com.slack.api:slack-app-backend:1.2.1
|         +--- com.slack.api:slack-api-model:1.2.1 (*)
|         +--- com.slack.api:slack-api-client:1.2.1 (*)
+--- com.slack.api:bolt-servlet:1.2.1
|    +--- com.slack.api:slack-api-model:1.2.1 (*)
|    +--- com.slack.api:slack-api-client:1.2.1 (*)
|    +--- com.slack.api:slack-app-backend:1.2.1 (*)
|    \--- com.slack.api:bolt:1.2.1 (*)
+--- com.slack.api:bolt:1.2.1
|    +--- com.slack.api:slack-api-model:1.2.1
|    +--- com.slack.api:slack-api-client:1.2.1
|    |    +--- com.slack.api:slack-api-model:1.2.1 (*)
|    \--- com.slack.api:slack-app-backend:1.2.1
|         +--- com.slack.api:slack-api-model:1.2.1 (*)
|         +--- com.slack.api:slack-api-client:1.2.1 (*)
+--- com.slack.api:bolt-servlet:1.2.1
|    +--- com.slack.api:slack-api-model:1.2.1 (*)
|    +--- com.slack.api:slack-api-client:1.2.1 (*)
|    +--- com.slack.api:slack-app-backend:1.2.1 (*)
|    \--- com.slack.api:bolt:1.2.1 (*)

#### Java Runtime version

openjdk version "11.0.9" 2020-10-20
OpenJDK Runtime Environment (build 11.0.9+11-Ubuntu-0ubuntu1.20.04)
OpenJDK 64-Bit Server VM (build 11.0.9+11-Ubuntu-0ubuntu1.20.04, mixed mode, sharing)

#### OS info

#57-Ubuntu SMP Thu Oct 15 10:57:00 UTC 2020

#### Steps to reproduce:

(Share the commands to run, source code, and project settings (e.g., pom.xml/build.gradle))

This application currently will not work unless vpn access is enabled due to an integration with a 3rd party product that is only available in a secured environment.  If this is a problem, I can isolate the issue into a separate code base and share an archive.

#### Expected result:

I'm trying to use [this](https://slack.com/oauth/v2/authorize?client_id=1413953711251.1437157832549&scope=im:history,users:read,chat:write&user_scope=dnd:read) Sharable URL to add a slack app I'm devloping to another workspace.  I'm actively working on the applications so that URL points to my development environment that is fronted by ngrok. 

#### Actual result:

I'm getting the following warning and then the flow is cancelled:
```
2020-10-29 09:11:42.620  WARN 173914 --- [nio-8080-exec-1] .b.s.b.o.d.OAuthDefaultStateErrorHandler : Invalid state parameter detected - payload: VerificationCodePayload(code=9225288469.1459344528309.4b4f4a7b0ed1b066294482e33cdcd4df0e6d874b1763cce87fb97c57f591a9b2, state=, error=null)
```

### Requirements

Please make sure if this topic is specific to this SDK. For general questions/issues about Slack API platform or its server-side, could you submit questions at https://my.slack.com/help/requests/new instead. :bow:

Please read the [Contributing guidelines](https://github.com/slackapi/java-slack-sdk/blob/main/.github/contributing.md) and [Code of Conduct](https://slackhq.github.io/code-of-conduct) before creating this issue or pull request. By submitting, you are agreeing to the those rules.

**Comment by seratch** [comment]
Hi @mattv thanks for asking the question!

The short answer is that Bolt apps do not use the Sharable URL in the Slack app configuration page as-is. You can use `SLACK_INSTALL_PATH` instead with a proper configuration.

You an extract `scope` parameter from the Sharable URL and set the value to `SLACK_SCOPES` env variable. 

Also, your OAuth flow should start from the URL that is set by `SLACK_INSTALL_PATH` - the default is https://{your domain}/slack/oauth/start` (we're going to recommend using `/slack/install` since the next major version though).

The `SLACK_INSTALL_PATH` endpoint generates the `https://slack.com/oauth/v2/authorize` URL with a valid `state` parameter for your app installers.

**Comment by mattv** [comment]
Thank you so much for your help.  I have been fighting this silly issue
since yesterday afternoon.  Once I started from the SLACK_INSTALL_PATH I
then discovered I had left off a character in my SLACK_SCOPES environment
variable value.  After I fixed that I was able to complete the OAuth flow.

Thanks again for your help.  I really appreciate it!

Regards,
    Matt Valerio
    valerio.matt@gmail.com
    614-432-4857
